src/internal/m365/collection/drive/delta_tree_test.go

```diff
@@ -651,3 +654,202 @@ func (suite *DeltaTreeUnitSuite) TestFolderyMcFolderFace_SetFolder_correctTombst
+
+func (suite *DeltaTreeUnitSuite) TestFolderyMcFolderFace_AddFile() {
+	treeWithFileAtRoot := func() *folderyMcFolderFace {
+		tree := treeWithRoot()
+		tree.root.files[id(file)] = time.Now()
+		tree.fileIDToParentID[id(file)] = rootID
+
+		return tree
+	}
+
+	treeWithFileInFolder := func() *folderyMcFolderFace {
+		tree := treeWithFolders()
+		tree.root.files[id(file)] = time.Now()
+		tree.fileIDToParentID[id(file)] = id(folder)
+
+		return tree
+	}
+
+	treeWithFileInTombstone := func() *folderyMcFolderFace {
+		tree := treeWithTombstone()
+		tree.tombstones[id(folder)].files[id(file)] = time.Now()
+		tree.fileIDToParentID[id(file)] = id(folder)
+
+		return tree
+	}
+
+	table := []struct {
+		tname       string
+		tree        *folderyMcFolderFace
+		oldParentID string
+		parentID    string
+		expectErr   assert.ErrorAssertionFunc
+		expectFiles map[string]string
+	}{
+		{
+			tname:       "add file to root",
+			tree:        treeWithRoot(),
+			oldParentID: "",
+			parentID:    rootID,
+			expectErr:   assert.NoError,
+			expectFiles: map[string]string{id(file): rootID},
+		},
+		{
+			tname:       "add file to folder",
+			tree:        treeWithFolders(),
+			oldParentID: "",
+			parentID:    id(folder),
+			expectErr:   assert.NoError,
+			expectFiles: map[string]string{id(file): id(folder)},
+		},
+		{
+			tname:       "re-add file at the same location",
+			tree:        treeWithFileAtRoot(),
+			oldParentID: rootID,
+			parentID:    rootID,
+			expectErr:   assert.NoError,
+			expectFiles: map[string]string{id(file): rootID},
+		},
+		{
+			tname:       "move file from folder to root",
+			tree:        treeWithFileInFolder(),
+			oldParentID: id(folder),
+			parentID:    rootID,
+			expectErr:   assert.NoError,
+			expectFiles: map[string]string{id(file): rootID},
+		},
+		{
+			tname:       "move file from tombstone to root",
+			tree:        treeWithFileInTombstone(),
+			oldParentID: id(folder),
+			parentID:    rootID,
+			expectErr:   assert.NoError,
+			expectFiles: map[string]string{id(file): rootID},
+		},
+		{
+			tname:       "error adding file to tombstone",
+			tree:        treeWithTombstone(),
+			oldParentID: "",
+			parentID:    id(folder),
+			expectErr:   assert.Error,
+			expectFiles: map[string]string{},
+		},
+		{
+			tname:       "error adding file before parent",
+			tree:        treeWithTombstone(),
+			oldParentID: "",
+			parentID:    idx(folder, 1),
+			expectErr:   assert.Error,
+			expectFiles: map[string]string{},
+		},
+	}
+	for _, test := range table {
+		suite.Run(test.tname, func() {
+			t := suite.T()
+
+			err := test.tree.addFile(
+				test.parentID,
+				id(file),
+				time.Now())
+			test.expectErr(t, err, clues.ToCore(err))
+			assert.Equal(t, test.expectFiles, test.tree.fileIDToParentID)
+
+			if err != nil {
+				return
+			}
+
+			parent, ok := test.tree.folderIDToNode[test.parentID]
+			if !ok {
+				parent = test.tree.tombstones[test.parentID]
+			}
```

Comment:
maybe a `getLiveOrDeadNode` function would be helpful? There's at least this location and a location when adding a file that has similar code

Reply:
sure, sounds reasonable